api/types/src/convert.rs

```diff
@@ -423,68 +422,114 @@ impl<'a, R: MoveResolver + ?Sized> MoveConverter<'a, R> {
-            | MoveType::Reference { mutable: _, to: _ } => {
-                return Err(format_err!(
-                    "unexpected move type {:?} for value {:?}",
-                    &typ,
-                    &val
-                ))
+        Ok(match layout {
+            MoveTypeLayout::Bool => Bool(serde_json::from_value::<bool>(val)?),
+            MoveTypeLayout::U8 => U8(serde_json::from_value::<u8>(val)?),
+            MoveTypeLayout::U64 => serde_json::from_value::<crate::U64>(val)?.into(),
+            MoveTypeLayout::U128 => serde_json::from_value::<crate::U128>(val)?.into(),
+            MoveTypeLayout::Address => serde_json::from_value::<crate::Address>(val)?.into(),
+            MoveTypeLayout::Vector(item_layout) => {
+                self.try_into_vm_value_vector(item_layout.as_ref(), val)?
+            }
+            MoveTypeLayout::Struct(struct_layout) => {
+                self.try_into_vm_value_struct(struct_layout, val)?
+            }
+            MoveTypeLayout::Signer => {
+                bail!("unexpected move type {:?} for value {:?}", layout, val)
             }
         })
     }
 
-    pub fn try_into_move_value_vector(
+    pub fn try_into_vm_value_vector(
         &self,
-        typ: &MoveType,
+        layout: &MoveTypeLayout,
         val: Value,
     ) -> Result<move_core_types::value::MoveValue> {
-        if matches!(typ, MoveType::U8) {
+        if matches!(layout, MoveTypeLayout::U8) {
             Ok(serde_json::from_value::<HexEncodedBytes>(val)?.into())
         } else if let Value::Array(list) = val {
             let vals = list
                 .into_iter()
-                .map(|v| self.try_into_move_value(typ, v))
+                .map(|v| self.try_into_vm_value_from_layout(layout, v))
                 .collect::<Result<_>>()?;
 
             Ok(move_core_types::value::MoveValue::Vector(vals))
         } else {
-            Err(format_err!(
-                "expected vector<{:?}>, but got: {:?}",
-                typ,
-                &val
-            ))
+            bail!("expected vector<{:?}>, but got: {:?}", layout, val)
         }
     }
 
+    pub fn try_into_vm_value_struct(
+        &self,
+        layout: &MoveStructLayout,
+        val: Value,
+    ) -> Result<move_core_types::value::MoveValue> {
+        let field_layouts = if let MoveStructLayout::WithFields(fields) = layout {
+            fields
+        } else {
+            bail!(
+                "Expecting `MoveStructLayout::WithFields, getting {:?}",
+                layout
+            );
+        };
+
+        let mut field_values = if let Value::Object(fields) = val {
+            fields
+        } else {
+            bail!("Expecting a JSON Map for struct.");
+        };
+
+        let fields = field_layouts
+            .iter()
+            .map(|field_layout| {
+                let name = field_layout.name.as_str();
+                let value = field_values
+                    .remove(name)
```

Comment:
Why do you need to remove it from the map? Can we not just do a `get` ?

Reply:
You get a ref from `get`, and you'll need to copy it.